src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,418 @@
+            loss, logits = self._evaluate_steps(features, labels)
+            loss = tf.reduce_mean(loss)
+
+            if not prediction_loss_only:
+                if self.args.n_gpu > 1:
+                    for val in logits.values:
+                        if preds is None:
+                            preds = val.numpy()
+                        else:
+                            preds = np.append(preds, val.numpy(), axis=0)
+
+                    for val in labels.values:
+                        if label_ids is None:
+                            label_ids = val.numpy()
+                        else:
+                            label_ids = np.append(label_ids, val.numpy(), axis=0)
+                else:
+                    if preds is None:
+                        preds = logits.numpy()
+                    else:
+                        preds = np.append(preds, logits.numpy(), axis=0)
+
+                    if label_ids is None:
+                        label_ids = labels.numpy()
+                    else:
+                        label_ids = np.append(label_ids, labels.numpy(), axis=0)
+
+            step += 1
+
+        if self.compute_metrics is not None and preds is not None and label_ids is not None:
+            metrics = self.compute_metrics(EvalPrediction(predictions=preds, label_ids=label_ids))
+        else:
+            metrics = {}
+
+        metrics["loss"] = loss.numpy()
+
+        return PredictionOutput(predictions=preds, label_ids=label_ids, metrics=metrics)
+
+    def evaluate(
+        self, eval_dataset: Optional[tf.data.Dataset] = None, prediction_loss_only: Optional[bool] = None
+    ) -> Dict[str, float]:
+        """
+        Prediction/evaluation loop, shared by `evaluate()` and `predict()`.
+        """
+        if eval_dataset is None:
+            eval_dataset = self.tf_eval_dataset
+
+        output = self._prediction_loop(eval_dataset, description="Evaluation")
+
+        return output.metrics
+
+    def train(self) -> None:
+        """
+        Train method to train the model.
+        """
+        tf.summary.trace_on(graph=True, profiler=True)
+        self.gradient_accumulator.reset()
+
+        iterations = self.optimizer.iterations
+        tf.summary.experimental.set_step(iterations)
+
+        epochs = 1 if self.args.max_steps > 0 else self.args.num_train_epochs
+
+        logger.info("***** Running training *****")
+        logger.info("  Num examples = %d", len(self.train_dataset))
+        logger.info("  Num Epochs = %d", self.args.num_train_epochs)
+        logger.info("  Total optimization steps = %d", self.train_steps)
+
+        for epoch in range(1, int(epochs + 1)):
+            for training_loss in self._training_steps():
+                step = iterations.numpy()
+                training_loss = tf.reduce_mean(training_loss)
+
+                with self.writer.as_default():
+                    tf.summary.scalar("loss", training_loss, step=step)
```

Comment:
here you're writing to TensorBoard at every step, right?

Reply:
Yes, I trace only the loss at every step.

Reply:
A single step's loss (with no averaging whatsoever) is kinda random though, no? 

Shouldn't we keep the same behaviour as in Trainer?

Reply:
I like to keep trace when I run a model for the first time to see how it evolves and check what are the important steps and then afterwards I can better tune the model, but I don't mind to remove it. As you wish 😄

Reply:
No I think we can keep it